### PR TITLE
feat: allow empty field mysql ssl [TCTC-5306]

### PR DESCRIPTION
**Pull request by Sanix-Darker**
## WHAT

we want to have ssl_ca, ssl_cert and ssl_key as optional params for an ssl connexion

### CODE REVIEW

File: toucan_connectors/mysql/mysql_connector.py [hunk 1]
```diff
@@ -231,25 +235,21 @@ def _connect(
...
-                    'ssl_ca': ssl_ca.name,
-                    'ssl_cert': ssl_cert.name,
-                    'ssl_key': ssl_key.name,
+                    **connection_params,
                     # Verify that the server's hostname matches the CA
-                    'ssl_verify_identity': self.ssl_mode == SSLMode.VERIFY_IDENTITY,
+                    **{
+                        'ssl_verify_identity': self.ssl_mode == SSLMode.VERIFY_IDENTITY
+                        or (getattr(self, 'ssl_key') is None or getattr(self, 'ssl_cert') is None)
+                    },
```
Comment (lukapeschke) [inline toucan_connectors/mysql/mysql_connector.py:252]: This beahviour is not valid: We only want to check if the hostname matches the cert if the mode is VERIFY_IDENTITY

File: toucan_connectors/mysql/mysql_connector.py [hunk 2]
```diff
@@ -231,25 +235,21 @@ def _connect(
...
-                # Writing cert contents to temporary files. They're opened in wb+ mode
-                ssl_ca.write(ssl_params['ssl_ca'].encode())
-                ssl_ca.seek(0)
-                ssl_cert.write(ssl_params['ssl_cert'].encode())
-                ssl_cert.seek(0)
-                ssl_key.write(ssl_params['ssl_key'].encode())
-                ssl_key.seek(0)
+            for ssl_opt in ('ssl_ca', 'ssl_key', 'ssl_cert'):
+                ssl_opt_type = NamedTemporaryFile(prefix=ssl_opt)
+                if ssl_opt in ssl_params:
```
Comment (lukapeschke) [inline toucan_connectors/mysql/mysql_connector.py:240]:
```suggestion
                if ssl_opt in ssl_params:
                    ssl_opt_file = NamedTemporaryFile(prefix=ssl_opt)
```

File: toucan_connectors/mysql/mysql_connector.py [hunk 3]
```diff
@@ -150,15 +150,19 @@ class Config:
...
         params = {}
         if self.ssl_mode in (SSLMode.VERIFY_CA, SSLMode.VERIFY_IDENTITY):
+
+            # if one is present, the other one should be available (should be
+            # better in a validator ?)
+            for k, p in (('ssl_key', 'ssl_cert'), ('ssl_cert', 'ssl_key')):
+                if getattr(self, k) is not None:
+                    assert (
+                        getattr(self, p) is not None
+                    ), f'SSL option {k} should be available if {p} is provided !'
```
Comment (lukapeschke) [inline toucan_connectors/mysql/mysql_connector.py:160]:
```suggestion
                    ), f'SSL option {k} should be specified if {p} is provided !'
```

File: toucan_connectors/mysql/mysql_connector.py [hunk 4]
```diff
@@ -231,25 +235,21 @@ def _connect(
...
-                ssl_cert.seek(0)
-                ssl_key.write(ssl_params['ssl_key'].encode())
-                ssl_key.seek(0)
+            for ssl_opt in ('ssl_ca', 'ssl_key', 'ssl_cert'):
+                ssl_opt_type = NamedTemporaryFile(prefix=ssl_opt)
+                if ssl_opt in ssl_params:
+                    ssl_opt_type.write(ssl_params[ssl_opt].encode())
+                    ssl_opt_type.seek(0)
+
+                    connection_params |= {**connection_params, **{ssl_opt: ssl_opt_type.name}}
```
Comment (lukapeschke) [inline toucan_connectors/mysql/mysql_connector.py:244]:
```suggestion
                    connection_params[ssl_opt] = ssl_opt_file.name
```

File: toucan_connectors/mysql/mysql_connector.py [hunk 5]
```diff
@@ -150,15 +150,19 @@ class Config:
     def _sanitize_ssl_params(self) -> dict[str, Any]:
         params = {}
         if self.ssl_mode in (SSLMode.VERIFY_CA, SSLMode.VERIFY_IDENTITY):
+
+            # if one is present, the other one should be available (should be
+            # better in a validator ?)
+            for k, p in (('ssl_key', 'ssl_cert'), ('ssl_cert', 'ssl_key')):
```
Comment (lukapeschke) [inline toucan_connectors/mysql/mysql_connector.py:156]: Sure, why not

**Review by lukapeschke — CHANGES_REQUESTED** [review]

File: tests/mysql/test_mysql.py [hunk 6]
```diff
@@ -477,13 +475,11 @@ def test_ssl_parameters_verify_identity(
     kwargs = connect_mock.call_args.kwargs
     assert kwargs['ssl_disabled'] is False
     assert kwargs['ssl_ca'] is not None
-    assert kwargs['ssl_cert'] is not None
-    assert kwargs['ssl_key'] is not None
```
Comment (lukapeschke) [inline tests/mysql/test_mysql.py:481]: Same remark here

File: toucan_connectors/mysql/mysql_connector.py [hunk 7]
```diff
@@ -231,25 +234,18 @@ def _connect(
...
 
                 connection_params |= {
-                    'ssl_ca': ssl_ca.name,
-                    'ssl_cert': ssl_cert.name,
-                    'ssl_key': ssl_key.name,
+                    **connection_params,
                     # Verify that the server's hostname matches the CA
-                    'ssl_verify_identity': self.ssl_mode == SSLMode.VERIFY_IDENTITY,
+                    **{'ssl_verify_identity': self.ssl_mode == SSLMode.VERIFY_IDENTITY},
                 }
```
Comment (lukapeschke) [inline toucan_connectors/mysql/mysql_connector.py:249]: This should be outside of the loop:

```python
connection_params["ssl_verify_identity"] = self.ssl_mode == SSLMode.VERIFY_IDENTITY}
```

**Review by lukapeschke — COMMENTED** [review]
A few inline comments. Also, have you been able to test this on a mysql server with mandatory SSL ?

File: toucan_connectors/mysql/mysql_connector.py (hunk 7 above)
Comment (Sanix-Darker) [reply toucan_connectors/mysql/mysql_connector.py:249]: 🤦🏽‍♂️

**Comment by Sanix-Darker** [comment]
Tested on a local stack:

- With CA certificate only GOOD/BAD (cred)
  - On a good/valid CA.pem:

Screenshot from 2023-02-21 21-14-10
  - On a Bad CA.pem:
  
Screenshot from 2023-02-21 21-15-10

- With Cert/Key one missing, we raise the appropriate validation error
Screenshot from 2023-02-21 20-33-53

**Review by lukapeschke — APPROVED** [review]

**Pull request merged**